File: src/catalog/manifest.ts
export const KNOWN_KINDS = ['command', 'stack'] as const;
export type KnownKind = (typeof KNOWN_KINDS)[number];

export interface ManifestEntry {
  kind: string;
  source: string;
}

export interface Manifest {
  version: string;
  requires_binary: string;
  entries: ManifestEntry[];
}

export function parseManifest(raw: unknown): Manifest {
  if (!raw || typeof raw !== 'object' || Array.isArray(raw)) {
    throw new Error('MANIFEST_PARSE_ERROR');
  }
  const obj = raw as Record<string, unknown>;
  if (typeof obj.version !== 'string') throw new Error('MANIFEST_PARSE_ERROR');
  if (typeof obj.requires_binary !== 'string') throw new Error('MANIFEST_PARSE_ERROR');
  if (!Array.isArray(obj.entries)) throw new Error('MANIFEST_PARSE_ERROR');

  const entries: ManifestEntry[] = obj.entries.map((e: unknown) => {
    if (!e || typeof e !== 'object' || Array.isArray(e)) throw new Error('MANIFEST_PARSE_ERROR');
    const entry = e as Record<string, unknown>;
    if (typeof entry.kind !== 'string') throw new Error('MANIFEST_PARSE_ERROR');
    if (typeof entry.source !== 'string') throw new Error('MANIFEST_PARSE_ERROR');
    return { kind: entry.kind, source: entry.source };
  });

  return { version: obj.version, requires_binary: obj.requires_binary, entries };
}

export function validateKinds(manifest: Manifest): { unknownKind: string } | null {
  for (const entry of manifest.entries) {
    if (!(KNOWN_KINDS as readonly string[]).includes(entry.kind)) {
      return { unknownKind: entry.kind };
    }
  }
  return null;
}

File: src/catalog/manifest.py
import json
import sys
from pathlib import Path

KNOWN_KINDS = {"command", "stack"}


def parse_manifest(path: Path) -> list:
    from src.errors import ManifestParseError

    try:
        content = path.read_text(encoding="utf-8")
    except (FileNotFoundError, OSError) as exc:
        raise ManifestParseError(f"No se encontró catalog.json en {path}") from exc

    try:
        data = json.loads(content)
    except json.JSONDecodeError as exc:
        raise ManifestParseError(f"catalog.json tiene JSON inválido: {exc}") from exc

    if not isinstance(data, dict):
        raise ManifestParseError("catalog.json no es un objeto JSON válido")

    raw_entries = data.get("entries", [])
    if not isinstance(raw_entries, list):
        raise ManifestParseError("El campo 'entries' de catalog.json debe ser una lista")

    entries = []
    for entry in raw_entries:
        kind = entry.get("kind", "")
        source = entry.get("source", "")

        if not source:
            raise ManifestParseError(
                f"Entrada en catalog.json sin campo 'source': {entry!r}"
            )

        if kind not in KNOWN_KINDS:
            print(
                f"WARN: Ignoré entrada con kind desconocido '{kind}' (source: {source}). "
                "Probablemente tu CLI está desactualizada; corré `from-scratch update`.",
                file=sys.stderr,
            )
            continue

        # Strip unknown fields, keep only what we need
        entries.append({"kind": kind, "source": source})

    return entries
